tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc

```diff
@@ -0,0 +1,514 @@
+  }
+
+  // Fusing too many computations at a time may not be easily profitable and
+  // may increase compile time due to large kernels. Set a limit to it.
+  constexpr int64 kMaxFusionBatchSize = 32;
+  // CUDA has a parameter size limit of ~4k bytes.
+  constexpr int64 kMaxCudaParamSize = 4000;
+  size_t accum_io_size = 0;
+  auto reach_max_fusion_batch_size = [&](size_t left, size_t right) -> bool {
+    if (right - left >= kMaxFusionBatchSize) {
+      return true;
+    }
+
+    accum_io_size += fusion_instrs_.at(right)->fused_parameters().size() +
+                     GetOutputSizeOfFusion(*fusion_instrs_.at(right));
+
+    if (accum_io_size * 8 >= kMaxCudaParamSize) {
+      return true;
+    }
+
+    return false;
+  };
+
+  size_t left = pos_;
+  size_t right = pos_ + 1;
+  size_t first_output_size = GetOutputSizeOfFusion(*fusion_instrs_[left]);
+  PrimitiveType first_output_type =
+      GetUniqueOutputTypeOfFusion(*fusion_instrs_[left]);
+  for (; right < fusion_instrs_.size(); ++right) {
+    PrimitiveType cur_output_type =
+        GetUniqueOutputTypeOfFusion(*fusion_instrs_[right]);
+    if (first_output_type != cur_output_type) {
+      // Cannot fuse computations who have multiple output types.
+      break;
+    } else if (first_output_size != GetOutputSizeOfFusion(*fusion_instrs_[right])) {
+      // Cannot fuse computations who have different numbers of outputs.
+      break;
+    } else if (fusion_instrs_[left]->fused_instruction_count() !=
+               fusion_instrs_[right]->fused_instruction_count()) {
+      // Do not fuse computations of different instruction counts as it may
+      // introduce control divergence. This is a very simple heuristic to avoid
+      // fusing computations with too much discrepancy and we may improve it
+      // when the needs arise.
+      break;
+    } else if (reach_max_fusion_batch_size(left, right)) {
+      // Hit max fusion batch size.
+      break;
+    }
+  }
+
+  pos_ = right;
+  return absl::MakeSpan(fusion_instrs_).subspan(left, right - left);
+}
+
+Status HorizontalFusionImpl::CreateFusedComputation(
+    absl::Span<HloInstruction*> fused_fusion_instrs,
+    std::unique_ptr<HloComputation>* uniq_computation,
+    std::vector<HloInstruction*>* bound_operands) {
+  // First, build a computation with only params.
+  HloComputation::Builder b("horizontally_fused_computation");
+  size_t fused_comp_param_id = 0;
+  for (size_t i = 0; i < fused_fusion_instrs.size(); ++i) {
+    auto old_params = fused_fusion_instrs[i]->fused_parameters();
+    for (size_t j = 0; j < old_params.size(); ++j) {
+      auto bound_opnd = fused_fusion_instrs[i]->mutable_operand(j);
+      // in a form of param_i_j
+      auto new_param = b.AddInstruction(HloInstruction::CreateParameter(
+          fused_comp_param_id++, bound_opnd->shape(),
+          absl::StrCat("param_", i, "_", j)));
+      bound_operands->push_back(bound_opnd);
+    }
+  }
+  *uniq_computation = b.Build();
```

Comment:
A few tests (e.g. `//tensorflow/compiler/xla/tests:prng_test_gpu`) seem to CHECK fail on this line.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/tests/BUILD#L1751

```
F0117 05:31:52.164265    7056 hlo_computation.cc:62] Check failed: nullptr != root ((null) vs. (null)) 
*** Check failure stack trace: ***
...
xla::HloComputation::Builder::Build()
xla::gpu::(anonymous namespace)::HorizontalFusionImpl::CreateFusedComputation()
xla::gpu::(anonymous namespace)::HorizontalFusionImpl::Fuse()
xla::gpu::(anonymous namespace)::HorizontalFusionImpl::Run()
...
```

Reply:
Fixed. It is a case where fusion instructions have no parameters at all. Then, no instructions was added into the HloComputation (and hence no roots) but existence of a root instruction is always required by this class.